tls/s2n_cipher_suites.c

```diff
@@ -1168,13 +1216,179 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
+ * @return true Cipher suite match is usable
+ * @return false Cipher suite match is not usable for this connection
+ */
+static bool s2n_cipher_suite_match_is_valid(struct s2n_connection* conn, struct s2n_cipher_suite* potential_match) {
+        /* Never use TLS1.3 ciphers on a pre-TLS1.3 connection, and vice versa */
+        if ((conn->actual_protocol_version >= S2N_TLS13) != (potential_match->minimum_required_tls_version >= S2N_TLS13)) {
+            return false;
+        }
+
+        /* Skip the suite if we don't have an available implementation */
+        if (!potential_match->available) {
+            return false;
+        }
+
+        /* Make sure the cipher is valid for available certs */
+        if (s2n_is_cipher_suite_valid_for_auth(conn, potential_match) != S2N_SUCCESS) {
+            return false;
+        }
+
+        /* TLS 1.3 does not include key exchange in cipher suites */
+        if (potential_match->minimum_required_tls_version < S2N_TLS13) {
+            /* If the kex is not supported continue to the next candidate */
+            bool kex_supported = false;
+            if (s2n_result_is_error(s2n_kex_supported(potential_match, conn, &kex_supported))) {
+                return false;
+            }
+
+            if (!kex_supported) {
+                return false;
+            }
+
+            /* If the kex is not configured correctly continue to the next candidate */
+            if (s2n_result_is_error(s2n_configure_kex(potential_match, conn))) {
+                return false;
+            }
+        }
+
+        /**
+        *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
+        *# The server MUST ensure that it selects a compatible PSK
+        *# (if any) and cipher suite.
+        **/
+        if (conn->psk_params.chosen_psk != NULL) {
+            if (potential_match->prf_alg != conn->psk_params.chosen_psk->hmac_alg) {
+                return false;
+            }
+        }
+
+        return true;
+}
+
+/**
+ * @brief Same as s2n_wire_ciphers_contain() except that it sets the match's index in the client wire to the index var.
+ * 
+ * @param match Cipher suite IANA which we wish to find a match for
+ * @param wire Client wire
+ * @param count Number of entries
+ * @param cipher_suite_len Length offset
+ * @return int Index of the match in the client wire. If a match was not found then -1 is returned.
+ */
+static int s2n_wire_ciphers_has_server_cipher_at(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
+{
+    int negative_one = -1;
+    for (uint32_t i = 0; i < count; i++) {
+        const uint8_t *theirs = wire + (i * cipher_suite_len) + (cipher_suite_len - S2N_TLS_CIPHER_SUITE_LEN);
+
+        if (!memcmp(match, theirs, S2N_TLS_CIPHER_SUITE_LEN)) {
+            return i;
+        }
+    }
+
+    return negative_one;
+}
+
+/**
+ * @brief This is the logic for parsing a list of s2n_cipher_suites* and selecting a cipher_suite accounting
```

Comment:
It would be helpful to have a comment explaining the basic algorithm that this function is implementing.

Also I don't really think you need to add a reference to documentation here. Since this function is internal to s2n comments should be directed to developers and not users.

Reply:
Will expand on the algorithm